### PR TITLE
Add testing flow to bootstrap secret storage

### PR DESCRIPTION
**Pull request by jryans**
This adds an testing button to the key backup panel which bootstraps the Secure
Secret Storage system (and also cross-signing keys).

2019-12-05 at 16 30
2019-12-05 at 16 31
2019-12-05 at 16 32

Uses https://github.com/matrix-org/matrix-js-sdk/pull/1079
Fixes https://github.com/vector-im/riot-web/issues/11212

### CODE REVIEW

**Comment by jryans** [comment]
CI failures look like package repo 502 errors, should be unrelated to the code.

**Comment by turt2live** [comment]
@jryans merging develop fixed the 502 problem, though the linter is unhappy as are the e2e tests.

File: src/MatrixClientPeg.js [hunk 1]
```diff
@@ -223,7 +223,7 @@ class MatrixClientPeg {
         if (SettingsStore.isFeatureEnabled("feature_cross_signing")) {
             // TODO: Cross-signing keys are temporarily in memory only. A
             // separate task in the cross-signing project will build from here.
```
Comment (dbkr) [inline src/MatrixClientPeg.js]: We should probably make sure we save them to SSSS in the same PR to save people from getting half setup cross-signing if they turn the labs flag on & push the button.

Comment (jryans) [reply src/MatrixClientPeg.js]: The JS SDK now saves them to SSSS when created. I have updated this comment to match, though it does make me wonder whether these callbacks are still needed.

File: src/async-components/views/dialogs/secretstorage/CreateSecretStorageDialog.js [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+                title={this._titleForPhase(this.state.phase)}
+                hasCancel={[PHASE_PASSPHRASE, PHASE_DONE].includes(this.state.phase)}
+            >
+            <div>
+                {content}
+            </div>
+            </BaseDialog>
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (jryans) [reply src/async-components/views/dialogs/secretstorage/CreateSecretStorageDialog.js]: Yes, mostly so... Copied key backup flows, then changed words and the actual "do it" bit.

File: src/components/views/dialogs/secretstorage/AccessSecretStorageDialog.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+limitations under the License.
+*/
+
+import React from 'react';
+import PropTypes from "prop-types";
+import sdk from '../../../../index';
+import MatrixClientPeg from '../../../../MatrixClientPeg';
+
+import { _t } from '../../../../languageHandler';
+import {Key} from "../../../../Keyboard";
```
[the remark this answers is not in the document]
Comment (jryans) [reply src/components/views/dialogs/secretstorage/AccessSecretStorageDialog.js]: Thanks, fixed.

File: src/CrossSigningManager.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+// This stores the secret storage private keys in memory for the JS SDK. This is
+// only meant to act as a cache to avoid prompting the user multiple times
+// during the same session. It is considered unsafe to persist this to normal
+// web storage. For platforms with a secure enclave, we will store this key
+// there.
+const secretStorageKeys = {};
+
+// XXX: This flow should maybe be reworked to allow retries in case of typos,
+// etc.
```
[the remark this answers is not in the document]
Comment (jryans) [reply src/CrossSigningManager.js]: Aha, didn't even notice that one! (I amended to the JS SDK side to [document](https://github.com/matrix-org/matrix-js-sdk/commit/14a2d7e8602917e28171fad008497c78643e4a14) it.)

[the remark this answers is not in the document]
Comment (jryans) [reply src/CrossSigningManager.js]: In the end, I needed a separate function, as the logic for signing vs. decryption keys is different.

File: src/CrossSigningManager.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+import Modal from './Modal';
+import sdk from './index';
+import { deriveKey } from 'matrix-js-sdk/lib/crypto/key_passphrase';
+import { decodeRecoveryKey } from 'matrix-js-sdk/lib/crypto/recoverykey';
+
+// This stores the cross-signing private keys in memory for the JS SDK. They are
+// also persisted to Secure Secret Storage in account data by the JS SDK when
+// created.
+const crossSigningKeys = {};
```
[the remark this answers is not in the document]
Comment (jryans) [reply src/CrossSigningManager.js]: Ah right, okay, I think I misunderstood that the we had expected the app to choose storing into SSSS via the callback. That does sound nice from a flexibility perspective, but I think I'll change the JS SDK to read / write SSSS by default and not rely on these callbacks, as it's quite difficult to get the order of operations correct here.

File: src/CrossSigningManager.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+// storage of these itself?
+export const getCrossSigningKey = k => crossSigningKeys[k];
+export const saveCrossSigningKeys = newKeys => Object.assign(crossSigningKeys, newKeys);
+
+// This stores the secret storage private keys in memory for the JS SDK. This is
+// only meant to act as a cache to avoid prompting the user multiple times
+// during the same session. It is considered unsafe to persist this to normal
+// web storage. For platforms with a secure enclave, we will store this key
+// there.
+const secretStorageKeys = {};
```
[the remark this answers is not in the document]
Comment (jryans) [reply src/CrossSigningManager.js]: Okay, let's trim some complexity out of this monster PR stack by removing the cache for now. Then we can get a feel for it as a team and adjust separately.

**Comment by jryans** [comment]
No worries! 😄 Thanks for the reviews!

**Pull request merged**